[2.9.0] #1151 Add documentation for OCI feature in Apps & Marketplace Section

martyav opened the pull request:
<!--
Check the Rancher docs issues to see if there is an existing issue for this pull request. If there is, enter the issue number below.
-->

Fixes #1151

## Reminders

- See the [README](../README.md) for more details on how to work with the Rancher docs.

- Verify if changes pertain to other versions of Rancher. If they do, finalize the edits on one version of the page, then apply the edits to the other versions.

- If the pull request is dependent on an upcoming release, make sure to target the release branch instead of `main`.

## Description

<!--
- What is the goal of this pull request? 
- What did you change? 
- Are there any other pull requests, tickets, or issues associated with this pull request?
-->

From the issue:

> - GitHub Issue - https://github.com/rancher/rancher/issues/29105
> - UI Issue - https://github.com/rancher/dashboard/issues/9815
> - GitHub PR - https://github.com/rancher/rancher/pull/43823
>
> Summary
> There is a new feature that is going to be released in the Apps and Marketplace category. That is support for OCI-based Helm repositories.
>
> Acceptance criteria
>
> 1. There are docs covering the functionality of the OCI Helm repository feature in Apps & Marketplace
> 1. There are docs covering debugging and logging for any known OCI Helm repository issues
> 1. There are docs covering the limitations of the feature.

## Comments

<!--
Any additional notes a reviewer should know before we review.
-->

martyav commented:
I'm going to open this up for review, but I'm relunctant to create the release branch yet, as 2.9 is several months from now. Added the DO NOT MERGE label and have to to remember to point this to the 2.9 branch at a more appropriate time.

martyav commented:
@rohitsakala 

I edited most of the draft provided, but the last section needs more instructions. 

> To view logs, enable the debug option of rancher 
>
> The first option if there is any discrepancy is to refresh a cluster repository 
> The last option is to delete the oci helm repository clusterrepo and readd it. This will not delete any already installed helm charts.

Where in the UI do they go to perform these steps?

Thank you!

martyav commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,80 @@
+
+The `spec.insecurePlainHttp` field allows insecure connections to OCI registries. When this field is set to `true`, Rancher will connect to the OCI endpoint without performing an SSL check. This works exactly the same as how the `spec.insecurePlainHttp` field works in [ORAS CLI](https://oras.land/docs/commands/use_oras_cli), since under the hood Rancher manager uses the ORAS library.
+
+The CRD that is linked to the OCI Helm registry is `ClusterRepo`.
+
+[Screenshot of the ClusterRepo YAML]<!-- Engineers - Can we get this screenshot? Thank you! -->
```
I left this part in, but it feels like we need a little more content here to explain the relevance of mentioning it in this section vs somewhere else, and what users can use that information for.

martyav commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,80 @@
+1. Enter a name and description for the registry. Select **OCI** as the target.
+1. Enter the URL for the registry. You can use this field to finetune how many charts you want to make availabe for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
+
+  1. **oci://\<registry-host\>/**: Every chart in the registry becomes available for installation, regardless of namespace or tag.
+  1. **oci://\<registry-host\>/\<namespace\>**: Every chart in the specified namespace within the registry becomes available for installation.
+  1. **oci://\<registry-host\>/\<namespace\>/\<chart-name\>**: Only the specified chart and any associated tags or versions of that chart become available for installation.
+  1. **oci://\<registry-host\>/\<namespace\>/\<chart-name\>:\<tag\>**: Only the chart with the specified tag becomes available for installation.
```
I opted not to format these as code snippets, so the backslashes are required as escape characters for the angle brackets.

martyav commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,80 @@
+
+The metadata field in index.yaml can store up to 30 MB of information about the Helm charts within an OCI-based registry.
+
+## Troubleshooting OCI-based Helm Registries <!-- Unedited draft -->
+
+To view logs, enable the debug option of rancher 
+
+The first option if there is any discrepancy is to refresh a cluster repository 
+The last option is to delete the oci helm repository clusterrepo and readd it. This will not delete any already installed helm charts.
```
I left this section unedited because it needs more descriptive instructions. I wasn't sure where in the UI to turn on debugging or which page in the docset we could point to for instructions.

rohitsakala commented:
@martyav Thank you for this PR. There has been some modification requested for this feature. So, I complete those modifications and get back to you.

martyav commented:
> @martyav Thank you for this PR. There has been some modification requested for this feature. So, I complete those modifications and get back to you.

@rohitsakala do you have any updates on this?

martyav commented:
@rohitsakala checking in again, do you have any updates?

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+---
+title: Using OCI Helm Chart Registries
```
Updated

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries"/>
+</head>
+
+Helm v3 introduced storing Helm charts as [Open Container Initiative (OCI)](https://opencontainers.org/about/overview/) artifacts in container registries. With Rancher v2.9.0, you can add [OCI-based Helm chart registries](https://helm.sh/docs/topics/registries/) alongside http- and Git-based repositories. This means you can deploy apps that are stored as OCI artifacts.
```
Updated to, "As of Rancher v2.9.0, this feature is experimental."

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+
+Helm v3 introduced storing Helm charts as [Open Container Initiative (OCI)](https://opencontainers.org/about/overview/) artifacts in container registries. With Rancher v2.9.0, you can add [OCI-based Helm chart registries](https://helm.sh/docs/topics/registries/) alongside http- and Git-based repositories. This means you can deploy apps that are stored as OCI artifacts.
+
+## Add an OCI Registry
```
Updated

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+3. In the left navigation bar, select **Apps > Repositories**.
+4. Click **Create**.
+5. Enter a **Name** and **Description** for the registry. Select **OCI** as the target.
+6. Enter the URL for the registry. The registry endpoint must not contain anything besides OCI Helm Chart artifacts. If you attempt to add an endpoint that contains any other kinds of files or artifacts, the OCI registry will not be added. 
```
Updated

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+3. In the left navigation bar, select **Apps > Repositories**.
+4. Click **Create**.
+5. Enter a **Name** and **Description** for the registry. Select **OCI** as the target.
+6. Enter the URL for the registry. The registry endpoint must not contain anything besides OCI Helm Chart artifacts. If you attempt to add an endpoint that contains any other kinds of files or artifacts, the OCI registry will not be added. 
```
Added, "The artifacts should all have unique names."

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+
+## Authentication for OCI Registries
+
+Rancher supports BasicAuth for OCI registries. You must create a [**BasicAuth** Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret). You can [create the secret through the Rancher UI](../kubernetes-resources-setup/secrets.md). 
```
Updated

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries"/>
+</head>
+
+Helm v3 introduced storing Helm charts as [Open Container Initiative (OCI)](https://opencontainers.org/about/overview/) artifacts in container registries. With Rancher v2.9.0, you can add [OCI-based Helm chart registries](https://helm.sh/docs/topics/registries/) alongside http- and Git-based repositories. This means that you can deploy apps that are stored as OCI artifacts. As of Rancher v2.9.0,  ...
```
In this case, the title of the page at https://helm.sh/docs/topics/registries/ is "Use OCI-based registries"

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+
+Rancher supports BasicAuth for OCI registries. You must create a [**BasicAuth** Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret). You can also [create the secret through the Rancher UI](../kubernetes-resources-setup/secrets.md). 
+
+Rancher 2.9.0 also adds the `spec.insecurePlainHttp` field, which allows insecure connections to OCI registries. When this field is set to `true`, Rancher connects to the OCI endpoint without performing an SSL check. This works exactly the same as how the `spec.insecurePlainHttp` field works in [ORAS CLI](https://oras.land/docs/commands/use_oras_cli), since Rancher uses the ORAS library.
```
Do you mean "Explanation of fields"? It seems like we should explain those fields in context with the steps they are related to. "There are more fields that the user can define". Where do these fields appear and how can the user define them? Do they all appear on the same screen where the user sets up authentication?

martyav replied:
Ok, I added the explanations as part of the step-by-step instructions, in order of where they appear in the UI

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+
+To view logs, enable the debug option of rancher 
+
+The first option if there is any discrepancy is to refresh a cluster repository 
```
https://github.com/rancher/rancher-docs/pull/1171#discussion_r1534122129

> I left this section unedited because it needs more descriptive instructions. I wasn't sure where in the UI to turn on debugging or which page in the docset we could point to for instructions.

martyav commented:
@rohitsakala I made a number of edits addressing suggestions. However, there are still three things unresolved:

1. The last section needs to describe how to enable the debug option
2. There are two sentences I wasn't sure how to edit and wrapped in HTML comment tags
3. The section explaining fields feels like it needs to be revised extensively, so that the fields are mentioned in context with where they can be found or used. Are they all on the Authentic part of the UI?

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,103 @@
+
+If you have a OCI-based Helm chart repository which doesn't implement the `Retry-After` header, or which behaves differently than Dockerhub, provide your response values with the `ExponentialBackOff` field. 
+
+For example, if you have an OCI-based Helm chart repository that doesn't send a HTTP header, but you know that it allows 50 requests in 24 hours, you can provide a `MinWait` value of 86400 seconds, a `MaxWait` of 90000 seconds, and a `Maxretry` of `1`. <!-- I'm not sure what these two sentences from the draft mean? --> This would allow adding the Helm repository complete but it would take 24 hour ...
```
Thanks! Updating and adding to next commit

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+1. Click **Apps > Charts**.
+1. Select the OCI repository from the dropdown.
+
+## Refresh an OCI Registry
```
Rate limiting section added

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,96 @@
+
+## Troubleshooting OCI-based Helm Registries <!-- Unedited draft -->
+
+To view logs, enable the debug option of rancher 
```
This section has now been edited and the link has been added

diogoasouza commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+1. Click **☰ > Cluster Management**.
+2. Find the name of the cluster whose repositories you want to access. Click **Explore** at the end of the cluster's row.
+3. In the left navigation bar, select **Apps > Repositories**.
+4. Click **Create**.
```
Can we add a screenshot here, similar to [this](https://ranchermanager.docs.rancher.com/v2.9/integrations-in-rancher/rancher-extensions) one in the extension doc in step 3.3?

martyav replied:
I don't have access to this version of Rancher I don't think, so we'd need some help to create the screenshot

diogoasouza commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+  
+  :::note
+  
+  You can use the **OCI URL** field to finetune how many charts from the registry are availabe for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
```
Do we need to add this section? AFAIK, this is standard for every oci registry, not something specific to Rancher. 
@rohitsakala keep me honest here

diogoasouza commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+7. Set up authentication. Select **Basicauth** from the authentication field and enter a username and password as required. Otherwise, create or select an **Authentication** secret. See [Authentication](#authentication-for-oci-based-helm-chart-repositories) for a full description.
+8. (optional) Enter a base64 encoded DER certificate in the **CA Cert Bundle** field. This field is for cases where you have a private OCI-based Helm chart repository and need Rancher to trust its certificates.   
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repsoitory, select **Insecure Plain Http**. See  [Authentication](#authentication-for-oci-based-helm-chart-repositories) for more details.
+10. (optional) If you know that your repository may respond with status code `429 Too Many Requests` (for example, if your repository is on DockerHub), fill out the fields under **Exponential Back Off**:
```
```suggestion
10. (optional) If your repository has a rate-limiting policy and may respond with status code `429 Too Many Requests` (for example, if your repository is on DockerHub), you may want to fill out the fields under **Exponential Back Off**:
```

diogoasouza commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+8. (optional) Enter a base64 encoded DER certificate in the **CA Cert Bundle** field. This field is for cases where you have a private OCI-based Helm chart repository and need Rancher to trust its certificates.   
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repsoitory, select **Insecure Plain Http**. See  [Authentication](#authentication-for-oci-based-helm-chart-repositories) for more details.
+10. (optional) If you know that your repository may respond with status code `429 Too Many Requests` (for example, if your repository is on DockerHub), fill out the fields under **Exponential Back Off**:
+  1. **Min Wait**: The default is 1 second.
```
```suggestion
  1. **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying.  The default is 1 second.
```

diogoasouza commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repsoitory, select **Insecure Plain Http**. See  [Authentication](#authentication-for-oci-based-helm-chart-repositories) for more details.
+10. (optional) If you know that your repository may respond with status code `429 Too Many Requests` (for example, if your repository is on DockerHub), fill out the fields under **Exponential Back Off**:
+  1. **Min Wait**: The default is 1 second.
+  1. **Max Wait**: The default is 1 second.
```
```suggestion
  1. **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying.  The default is 5 second.
```

The default value is 5 seconds

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries.md:
```diff
@@ -0,0 +1,107 @@
+
+7. Set up authentication. Select **Basicauth** from the authentication field and enter a username and password as required. Otherwise, create or select an **Authentication** secret. See [Authentication](#authentication-for-oci-based-helm-chart-repositories) for a full description.
+8. (optional) Enter a base64 encoded DER certificate in the **CA Cert Bundle** field. This field is for cases where you have a private OCI-based Helm chart repository and need Rancher to trust its certificates.   
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repsoitory, select **Insecure Plain Http**. See  [Authentication](#authentication-for-oci-based-helm-chart-repositories) for more details.
```
```suggestion
9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repository, select **Insecure Plain Http**.
```

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+
+Dockerhub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
+
+Rancher currently checks for the `Retry-After` header. It also handles Dockerhub-style responses and automatically waits before making a new request. When handling Dockerhub-style responses, Rancher ignores `ExponentialBackOff` values. 
```
@rohitsakala or @diogoasouza  I had originally written, "When handling Dockerhub-style responses, Rancher ignores `ExponentialBackOff` values." but that must be a misunderstanding. How does the Dockerhub-specific behavior of a 6 hour wait get triggered vs Exponential Back Off values being used for other servers that use the same status code?

martyav replied:
Ok, I'm about to push changes which clarify this point

martyav commented:
@btat @LucasSaintarbor @sunilarjun Could you have a look? This has already been reviewed by SMEs and gained approval

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+8. (optional) Enter a base64 encoded DER certificate in the **CA Cert Bundle** field. This field is for cases where you have a private OCI-based Helm chart repository and need Rancher to trust its certificates.   
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repository, select **Insecure Plain Http**.
+10. (optional) If your repository has a rate-limiting policy and may respond with status code `429 Too Many Requests`, you may want to fill out the fields under **Exponential Back Off**:
+  1. **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying.  The default is 1 second.
```
```suggestion
    - **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying. The default is 1 second.
```

sunilarjun replied:
There was an issue with displaying the numbered list for step 10 correctly, so making a suggestion to move to a bulleted list for these items.

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repository, select **Insecure Plain Http**.
+10. (optional) If your repository has a rate-limiting policy and may respond with status code `429 Too Many Requests`, you may want to fill out the fields under **Exponential Back Off**:
+  1. **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying.  The default is 1 second.
+  1. **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying.  The default is 5 second.
```
```suggestion
    - **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying. The default is 5 second.
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+10. (optional) If your repository has a rate-limiting policy and may respond with status code `429 Too Many Requests`, you may want to fill out the fields under **Exponential Back Off**:
+  1. **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying.  The default is 1 second.
+  1. **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying.  The default is 5 second.
+  1. **Max Number of Retries**: The default is 5 retries.
```
```suggestion
    - **Max Number of Retries**: The default is 5 retries.
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+  1. **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying.  The default is 1 second.
+  1. **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying.  The default is 5 second.
+  1. **Max Number of Retries**: The default is 5 retries.
+Once these values are set, Rancher responds to the `429` status code by staggering requests based on the minimum and maximum wait values. The wait time between retries increases exponentially, until Rancher has sent the maximum number of retries set. See [Rate Limiting](#rate-limiting-of-oci-based-helm-chart-repositories) for more details.
```
```suggestion

    Once these values are set, Rancher responds to the `429` status code by staggering requests based on the minimum and maximum wait values. The wait time between retries increases exponentially, until Rancher has sent the maximum number of retries set. See [Rate Limiting](#rate-limiting-of-oci-based-helm-chart-repositories) for more details.
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+
+If you need to update immediately, you can [perform a manual refresh](../helm-charts-in-rancher/helm-charts-in-rancher.md#refresh-chart-repositories).
+
+## Update a OCI-Based Helm Chart Repository Configuration
```
```suggestion
## Update an OCI-Based Helm Chart Repository Configuration
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+
+Rancher currently checks for the `Retry-After` header. It also handles Dockerhub-style responses (status code `429` and the `RateLimit-Remaining` header) and automatically waits before making a new request. When handling `Retry-After` or Dockerhub-style responses, Rancher ignores `ExponentialBackOff` values. 
+
+If you have a OCI-based Helm chart repository which doesn't implement the `Retry-After` or `RateLimit-Remaining` headers, and think you may be rate-limited at some point, fill out the fields under **Exponential Back Off** when you add the repository. 
```
```suggestion
If you have an OCI-based Helm chart repository which doesn't implement the `Retry-After` or `RateLimit-Remaining` headers, and think you may be rate-limited at some point, fill out the fields under **Exponential Back Off** when you add the repository. 
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+
+If you have a OCI-based Helm chart repository which doesn't implement the `Retry-After` or `RateLimit-Remaining` headers, and think you may be rate-limited at some point, fill out the fields under **Exponential Back Off** when you add the repository. 
+
+For example, if you have an OCI-based Helm chart repository that doesn't return a `Retry-After` header, but you know that the server allows 50 requests in 24 hours, you can provide Rancher a **Min Wait** value of 86400 seconds, a **Max Wait** value of 90000 seconds, and a **Max Number of Retries** value of **1**. Then, if Rancher gets rate limited by the server, Rancher will wait for 24 hours bef ...
```
```suggestion
For example, if you have an OCI-based Helm chart repository that doesn't return a `Retry-After` header, but you know that the server allows 50 requests in 24 hours, you can provide Rancher a **Min Wait** value of **86400** seconds, a **Max Wait** value of **90000** seconds, and a **Max Number of Retries** value of **1**. Then, if Rancher gets rate limited by the server, Rancher will wait for 24 hours before trying again. The request should succeed as Rancher hasn't sent any other requests in the previous 24 hours.
```

sunilarjun replied:
Matching the styling of bolding the 1 value for users.

sunilarjun reviewed (changes requested):
Hey Marty, left some review edits for some structural updates as well as some minor syntax/grammar updates. 

Also the sidebar navigation has to be updated to display the page (ie adding `"how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories"` to the `version-2.9-sidebars.json` file.

Also I think these pages should also be placed in the /docs folder for latest as the changes apply there as well and need to be added to the sidebars.json file too.

Thanks for handling Marty!

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/helm-charts-in-rancher.md:
```diff
@@ -163,6 +163,10 @@ spec:
 
+### Add Custom OCI Chart Repositories
+
+Helm v3 introduced storing Helm charts as [Open Container Initiative (OCI)](https://opencontainers.org/about/overview/) artifacts in container registries. With Rancher v2.9.0, you can add [OCI-based Helm chart repositories](https://helm.sh/docs/topics/registries/) alongside HTTP-based and Git-based repositories. This means you can deploy apps that are stored as OCI artifacts. For more information ...
```
@diogoasouza or @rohitsakala do you think it merits a caution note?

martyav commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,106 @@
+  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-registries"/>
+</head>
+
+Helm v3 introduced storing Helm charts as [Open Container Initiative (OCI)](https://opencontainers.org/about/overview/) artifacts in container registries. With Rancher v2.9.0, you can add [OCI-based Helm chart repositories](https://helm.sh/docs/topics/registries/) alongside HTTP-based and Git-based repositories. This means that you can deploy apps that are stored as OCI artifacts. As of Rancher v ...
```
I guess there's also the question of what experimental really means in this context. For feature flags, "experimental" means "not on by default" or "considered optional".

btat replied:
> I guess there's also the question of what experimental really means in this context. 

Features classified as "experimental" are unsupported (which requires a prominent callout) and often serve as a "preview" of upcoming features.

> For feature flags, "experimental" means "not on by default" or "considered optional".

I don't think the context of being behind a feature flag affects the meaning of "experimental" either. Because of their unsupported nature, experimental features are "considered optional" regardless of whether they're behind a feature flag or not. The "not on by default" part is not accurate either as we've had things like RKE2/K3s provisioning being on by default in early 2.6, but it was experimental.

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+9. (optional) To allow insecure connections without performing an SSL check, select **Skip TLS Verification**. To force Rancher to use HTTP instead of HTTPS to send requests to the repository, select **Insecure Plain Http**.
+10. (optional) If your repository has a rate-limiting policy and may respond with status code `429 Too Many Requests`, you may want to fill out the fields under **Exponential Back Off**:
+    - **Min Wait**: The minimum duration in seconds that Rancher should wait before retrying. The default is 1 second.
+    - **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying. The default is 5 second.
```
```suggestion
    - **Max Wait**: The maximum duration in seconds that Rancher should wait before retrying. The default is 5 seconds.
```

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+    - **Max Number of Retries**: The default is 5 retries.
+
+    Once these values are set, Rancher responds to the `429` status code by staggering requests based on the minimum and maximum wait values. The wait time between retries increases exponentially, until Rancher has sent the maximum number of retries set. See [Rate Limiting](#rate-limiting-of-oci-based-helm-chart-repositories) for more details.
+11. Add any labels and annotations.
```
```suggestion
11. (optional) Add any labels and annotations.
```

Adding `optional` based on the phrasing used. If it's not optional, we should specify what labels and annotations need to be added.

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+## Authentication for OCI-Based Helm Chart Repositories
+
+Rancher supports BasicAuth for OCI registries. You must create a [**BasicAuth** Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret). You can also [create the secret through the Rancher UI](../kubernetes-resources-setup/secrets.md). 
+
```
```suggestion
```

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+
+Most servers return a `Retry-After` header, indicating how long to wait before rate-limiting is lifted. 
+
+Dockerhub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
```
```suggestion
Docker Hub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
```

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+
+Dockerhub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
+
+Rancher currently checks for the `Retry-After` header. It also handles Dockerhub-style responses (status code `429` and the `RateLimit-Remaining` header) and automatically waits before making a new request. When handling `Retry-After` or Dockerhub-style responses, Rancher ignores `ExponentialBackOff` values. 
```
```suggestion
Rancher currently checks for the `Retry-After` header. It also handles Docker Hub-style responses (status code `429` and the `RateLimit-Remaining` header) and automatically waits before making a new request. When handling `Retry-After` or Docker Hub-style responses, Rancher ignores `ExponentialBackOff` values. 
```

btat commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+
+To enhance logging information, [enable the debug option](../../../troubleshooting/other-troubleshooting-tips/logging.md#kubernetes-install) while deploying Rancher.
+
+If there is any discrepancy between the repository contents and Rancher, you should refresh the cluster repository as a first resort. If the discrepancy persists and you've exhausted your options, delete the OCI-based Helm chart repository from Rancher and add it again. Deleting the repository won't delete any Helm charts that are already installed. 
```
`and you've exhausted your options` - what are other possible options?

martyav replied:
@rohitsakala are there other options we should mention/link?

martyav replied:
I'm just going to remove "and you've exhausted your options" until we can offer any options

btat reviewed (changes requested):
There are occurrences of both `rate limiting` and `rate-limiting` being used (often in the same context).

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+
+:::caution
+
+This feature is experimental.
```
Considering [our own team's uncertainty](https://github.com/rancher/rancher-docs/pull/1171#discussion_r1674511021) around the term, we should probably expand what it means to be experimental similar to https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/rancher-on-arm64 (although the recommendation may be excessive for this scenario).

martyav replied:
Maybe just "[x] is currently an experimental feature and is not yet officially supported in Rancher. "

btat replied:
> Maybe just "[x] is currently an experimental feature and is not yet officially supported in Rancher. "

I think that's fine if we drop the `yet` from it.

btat commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,113 @@
+  
+  You can use the **OCI URL** field to fine-tune how many charts from the registry are availabe for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
+
+    - **oci://\<registry-host\>/**: Every chart in the registry becomes available for installation, regardless of namespace or tag.
+    - **oci://\<registry-host\>/\<namespace\>**: Every chart in the specified namespace within the registry becomes available for installation.
+    - **oci://\<registry-host\>/\<namespace\>/\<chart-name\>**: Only the specified chart and any associated tags or versions of that chart become available for installation.
+    - **oci://\<registry-host\>/\<namespace\>/\<chart-name\>:\<tag\>**: Only the chart with the specified tag becomes available for installation.
```
This syntax looks fine, but when I preview via `yarn start` I'm seeing the escape characters getting rendered.

![pr1171](https://github.com/user-attachments/assets/a906b688-32eb-439f-a6c9-00a02eeb3512)

martyav replied:
updated the formatting to use backticks instead

btat commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,115 @@
+
+Most servers return a `Retry-After` header, indicating how long to wait before rate limiting is lifted. 
+
+Dockerhub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
```
```suggestion
Docker Hub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
```

btat commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,115 @@
+
+Dockerhub returns a `429` status code when it completes all allocated requests. It also returns a `RateLimit-Remaining` header which describes the rate limiting policy. 
+
+Rancher currently checks for the `Retry-After` header. It also handles Dockerhub-style responses (status code `429` and the `RateLimit-Remaining` header) and automatically waits before making a new request. When handling `Retry-After` or Dockerhub-style responses, Rancher ignores `ExponentialBackOff` values. 
```
```suggestion
Rancher currently checks for the `Retry-After` header. It also handles Docker Hub-style responses (status code `429` and the `RateLimit-Remaining` header) and automatically waits before making a new request. When handling `Retry-After` or Docker Hub-style responses, Rancher ignores `ExponentialBackOff` values. 
```

sunilarjun commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,115 @@
+  
+  :::note
+  
+  You can use the **OCI URL** field to fine-tune how many charts from the registry are availabe for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
```
```suggestion
  You can use the **OCI URL** field to fine-tune how many charts from the registry are available for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,114 @@
+  
+  :::note
+  
+  You can use the **OCI URL** field to fine-tune how many charts from the registry are availabe for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
```
```suggestion
  You can use the **OCI URL** field to fine-tune how many charts from the registry are available for installation on Rancher. More generic endpoints target more charts, as the following examples demonstrate:
```

sunilarjun commented on versioned_docs/version-2.9/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,114 @@
+
+- If there is any discrepancy between the repository contents and Rancher, you should refresh the cluster repository as a first resort. If the discrepancy persists, delete the OCI-based Helm chart repository from Rancher and add it again. Deleting the repository won't delete any Helm charts that are already installed.
+
+- Apps installed through OCI-based Helm chart repositories are subject to a known issue with how Rancher displays upgradeable version information. See the [Limitations](./helm-charts-in-rancher.md#limitations) section of **Helm Charts and Apps** for more deatils.  
```
```suggestion
- Apps installed through OCI-based Helm chart repositories are subject to a known issue with how Rancher displays upgradeable version information. See the [Limitations](./helm-charts-in-rancher.md#limitations) section of **Helm Charts and Apps** for more details.  
```

sunilarjun commented on docs/how-to-guides/new-user-guides/helm-charts-in-rancher/oci-repositories.md:
```diff
@@ -0,0 +1,115 @@
+
+- If there is any discrepancy between the repository contents and Rancher, you should refresh the cluster repository as a first resort. If the discrepancy persists, delete the OCI-based Helm chart repository from Rancher and add it again. Deleting the repository won't delete any Helm charts that are already installed.
+
+- Apps installed through OCI-based Helm chart repositories are subject to a known issue with how Rancher displays upgradeable version information. See the [Limitations](./helm-charts-in-rancher.md#limitations) section of **Helm Charts and Apps** for more deatils.
```
```suggestion
- Apps installed through OCI-based Helm chart repositories are subject to a known issue with how Rancher displays upgradeable version information. See the [Limitations](./helm-charts-in-rancher.md#limitations) section of **Helm Charts and Apps** for more details.
```

sunilarjun reviewed (changes requested):
Just some spelling updates, otherwise looks good!
